Core/Systems/CombatMountSystem/CombatMountPlayer.cs

```diff
@@ -0,0 +1,118 @@
+﻿using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Terraria;
+using Terraria.ModLoader;
+
+namespace StarlightRiver.Core.Systems.CombatMountSystem
+{
+	internal class CombatMountPlayer : ModPlayer
+	{
+		public CombatMount activeMount;
+
+		public int mountingTime;
+
+		public override void PreUpdateMovement()
+		{
+			if (Player.mount is null || !Player.mount.Active)
+				Dismount();
+
+			if (activeMount is null)
+				return;
+
+			if (activeMount.primaryCooldownTimer > 0) activeMount.primaryCooldownTimer --;
+			if (activeMount.secondaryCooldownTimer > 0) activeMount.secondaryCooldownTimer--;
+
+			if (activeMount.primaryAttackTimer > 0)
+			{
+				activeMount.primaryAttackTimer--;
+				activeMount.PrimaryAction(activeMount.primaryAttackTimer, Player);
+			}
+
+			if (activeMount.secondaryAbilityTimer > 0)
+			{
+				activeMount.secondaryAbilityTimer--;
+				activeMount.SecondaryAction(activeMount.secondaryAbilityTimer, Player);
+			}
+
+			activeMount.UpdatePhysics(Player);
+		}
+
+		public void Dismount()
+		{
+			activeMount = null;
+			Player.mount.Dismount(Player);
+		}
+	}
+
+	internal class CombatMountGlobalItem : GlobalItem
+	{
+		public override bool PreDrawInInventory(Item item, SpriteBatch spriteBatch, Vector2 position, Rectangle frame, Color drawColor, Color itemColor, Vector2 origin, float scale)
+		{
+			var isValid = ((item.DamageType.Type == DamageClass.Summon.Type || item.DamageType.Type == DamageClass.SummonMeleeSpeed.Type) && !Main.LocalPlayer.controlSmart);
+
+			if (!Main.playerInventory && !isValid && Main.LocalPlayer.GetModPlayer<CombatMountPlayer>().activeMount != null)
+			{
+				var tex = Terraria.GameContent.TextureAssets.Item[item.type].Value;
+				spriteBatch.Draw(tex, position, frame, drawColor * 0.25f, 0, origin, scale, 0, 0);
+				return false;
+			}
+
+			return true;
+		}
+
+		public override bool? CanAutoReuseItem(Item item, Player player)
+		{
+			var activeMount = player.GetModPlayer<CombatMountPlayer>().activeMount;
+
+			if (activeMount != null)
+				return activeMount.autoReuse;
+
+			return null;
+		}
+
+		public override bool AltFunctionUse(Item item, Player player)
+		{
+			var activeMount = player.GetModPlayer<CombatMountPlayer>().activeMount;
+
+			if (activeMount != null)
+				return true;
+
+			return false;
+		}
+
+		public override bool CanUseItem(Item item, Player player)
```

Comment:
this shouldn't be calling the start action functions and only decide if an item can be used or not -- which this probably should call a function on the mount from here since mounts might end up having differing restrictions. 

instead the mounts should perform their weapon logic every frame with a controls packet and decide whether or not to charge up to fire or not, this would allow for things like burst fires to go through even after releasing the button, or for charge shots to visually and mechanically cool down if you charged them but didn't go all the way before shooting.

Reply:
This creates alot of issues with vanilla input handling. It would require manually reimplementing all vanilla checks for if a mouse action can fire and would be incompatible with other mods attempting to cancel item usability. I had actually implemented it this way initially, but changed to this due to a number of issues with inputs firing at the wrong times.